Update helm template -x to support children chart manifests in more cases

chancez opened the pull request:
The previous code validated the manifests exist within the file system,
which only works when a chart is unpacked, meaning children charts
stored as tgz's in the parent chart's `charts/` directory don't exist as
individual files in the filesystem, causing validation to file.

Instead, we validate against the paths within the chart itself rather
than the filesystem, handling both chart.tgz's and charts that are
unpacked as directories in the chart/ directory.

A simple example/use-case of where this previous failed, but does not after my change:

helm-operator is a dependency of chargeback-helm-operator, in my requirements.yaml. The `helm-operator.tgz` is obtained from `helm dep build/update`

```
$ tree chargeback-helm-operator
chargeback-helm-operator
├── Chart.yaml
├── charts
│   └── helm-operator-0.1.0.tgz
├── requirements.lock
├── requirements.yaml
└── values.yaml

1 directory, 5 files
```

And the helm-operator chart:
```
helm-operator
├── Chart.yaml
├── templates
│   ├── cr.yaml
│   ├── crd.yaml
│   ├── deployment.yaml
│   ├── role.yaml
│   ├── rolebinding.yaml
│   └── service-account.yaml
└── values.yaml

1 directory, 8 files
```

Given the above, the following would fail:
```
helm template charts/chargeback-helm-operator -x charts/helm-operator/templates/deployment.yaml
Error: could not resolve template path: stat /Users/chance/go/src/github.com/coreos-inc/kube-chargeback/charts/chargeback-helm-operator/charts/helm-operator/templates/deployment.yaml: no such file or directory
```

With this PRs changes the same command renders the `deployment.yaml` manifest correctly.

Commit:
Update helm template -x to support children chart manifests in more cases

The previous code validated the manifests exist within the file system,
which only works when a chart is unpacked, meaning children charts
stored as tgz's in the parent chart's `charts/` directory don't exist as
individual files in the filesystem, causing validation to file.

Instead, we validate against the paths within the chart itself rather
than the filesystem, handling both chart.tgz's and charts that are
unpacked as directories in the chart/ directory.

chancez commented:
I took some liberties in refactoring/ordering the code since it was terribly difficult to understand before, as there was little comments and variable names weren't very descriptive/discernible.

thomastaylor312 reviewed (changes requested):
@chancez Just one concern about how to handle non-existent paths. I tested out the code and if I give a bogus path, it returns 0 with no rendered manifests instead of an error. So it looks like something isn't catching things properly

bacongobbler commented:
Perhaps some unit tests to cover these scenarios would be a good next approach :)

chancez commented:
Oh..hm I had made some changes right before opening my PR that may have removed the error handling I did have. Let me revisit that.

Commit:
Add working template to mariner chart for use in testing subcharts

Commit:
test helm template -x against subcharts stored as tgz's

Commit:
Disambigutate chartPaths variables in helm template tests

chancez commented:
@bacongobbler @thomastaylor312 I added some tests, and realized they actually still worked on master, then recalled my case was specifically with subcharts that are .tgzs, so I updated some of the testdata to support testing this use-case, and added a test that uses a subchart stored in a .tgz.

Commit:
Correctly use subtests in helm template tests

The testing.T in the subtest wasn't being used, so subtests were
effectively not being used. Additionally, in case subtests run in
parallel, we assign tt := tt to fix closure captures on range variables,
since the variables produced by a range are re-used between iterations.

Commit:
test helm template -x with non-existent manifest

chancez commented:
Also found another issue in how the tests were using subtests, so I fixed that. I also realized there were no tests that handle error cases, so I added a test that validates an error is thrown when passing a non-existent template to `-x`.

thomastaylor312 commented on pkg/chartutil/testdata/mariner/templates/placeholder.tpl:
```diff
@@ -1 +1,3 @@
 # This is a placeholder.
+
+Goodbye {{.Values.name | default "world"}}
```
💯

thomastaylor312 commented:
@chancez Thanks for adding (and fixing) the unit tests on this!

thomastaylor312 commented:
Waiting on second review now that this is a size large

SlickNik reviewed (approved):
Reviewed this and did a quick smoke test as well. Thanks for the fix @chancez! LGTM 👍
